### PR TITLE
[GEOT-7547]:Upgrading NetCDF version to 5.5.3

### PR DESCRIPTION
**Pull request by dromagnoli**
[GEOT-7547](https://osgeo-org.atlassian.net/browse/GEOT-7547) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

The downstream GeoServer integration build may fail since it requires the same upgrade.
The relevant GeoServer PR is https://github.com/geoserver/geoserver/pull/7507

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]
Good, overall less changes than I thought I would see. Maybe have some documentation in the ``upgrade.rst`` file here as well? No need to tell users how to perform upgrades, but make them aware that some breakage is possible with an upgrade, and point to the details in the Jira ticket?

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/utilities/NetCDFUtilities.java [hunk 1]
```diff
@@ -476,17 +446,63 @@ public enum FileFormat {
...
+            int period =
+                    Integer.getInteger(
+                            "org.geotools.coverage.io.netcdf.cache.cleanup.period", 12 * 60);
+
+            // No replacements found for 5.5 even if deprecated. Documentation still uses it
+            // https://docs.unidata.ucar.edu/netcdf-java/5.5/userguide/disk_caching.html#netcdffilecache
+
+            NetcdfDatasets.initNetcdfFileCache(minElements, maxElements, period);
+            ucar.unidata.io.RandomAccessFile.setGlobalFileCache(
+                    new FileCache(minElements, maxElements, period));
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/utilities/NetCDFUtilities.java:497]: Nevermind, it was there at the beginning when using the NetcdfDataset class. (without the final "s").
When using methods from NetcdfDatasets (with final "s") the deprecation went away but I didn't removed the annotation so I will remove it.

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/utilities/NetCDFUtilities.java [hunk 2]
```diff
@@ -615,6 +626,7 @@ public static boolean isVariableAccepted(final Variable var, final CheckType che
      * NetCDF files may contains a wide set of coverageDescriptorsCache. Some of them are unuseful
      * for our purposes. The method returns {@code true} if the specified variable is accepted.
      */
+    @SuppressWarnings("deprecation") // getFullName has no alternative
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/utilities/NetCDFUtilities.java:629]: See previous comment on Dimension.getFullName. It has no alternative

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/utilities/NetCDFUtilities.java [hunk 3]
```diff
@@ -1131,13 +1158,15 @@ public static void addIgnoredDimension(String dimensionName) {
      * @param var Variable instance
      * @return a Number representing NoData
      */
+    @SuppressWarnings("deprecation") // getFullName has no alternative
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/utilities/NetCDFUtilities.java:1161]: Same as above, that's a Dimension, not a Variable.

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java [hunk 4]
```diff
@@ -828,7 +828,7 @@ private void initSpatialDomain() throws Exception {
         spatialDomain.setGridGeometry(getGridGeometry());
     }
 
-    /** */
+    @SuppressWarnings("deprecation") // No alternatives for getFullName
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:831]: same as above, about Dimension.

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java [hunk 5]
```diff
@@ -1150,6 +1150,7 @@ public int[] splitIndex(int imageIndex) {
         return resultIndex;
     }
 
+    @SuppressWarnings("deprecation") // No alternative for getFullName
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:1153]: Dimension too

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java [hunk 6]
```diff
@@ -1320,7 +1321,7 @@ private String getTimeAttribute(CoordinateSystem cs) {
      * @param dimensionIndex the index of the dimension
      * @return the value
      */
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "deprecation"}) // No alternative for getFullName
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:1324]: Dimension

File: modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/CoordinateVariableTest.java [hunk 7]
```diff
@@ -70,6 +63,7 @@ public String getUnitsString() {
             return units;
         }
 
+        @SuppressWarnings("deprecation") // No alternative provided
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/CoordinateVariableTest.java:66]: Well, we are not actually constructing a new CoordinateSystem. 
We are wrapping the existing one into a new CoordinateSystemAdapter so I think that we can keep the deprecation but with a different comment. What to you think?

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java [hunk 8]
```diff
@@ -1342,6 +1343,7 @@ static class CoordinateSystemAdapter extends CoordinateSystem {
         /** Boolean indicating that the vertical axis is present */
         private final boolean vertical;
 
+        @SuppressWarnings("deprecation") // No alternative for CoordinateSystem
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:1346]: Well, we are not actually constructing a new CoordinateSystem. 
We are wrapping the existing one into a new CoordinateSystemAdapter so I think that we can keep the deprecation but with a different comment. What to you think?

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java [hunk 9]
```diff
@@ -488,12 +489,11 @@ public int getRank() {
...
     }
 
-    /** @throws Exception */
-    private void initSlicesInfo() throws Exception {
+    private void initSlicesInfo() {
         int[] shape = variableDS.getShape();
         numberOfSlices = 1;
         for (int i = 0; i < variableDS.getShape().length - 2; i++) {
-            if (!ignoredDimensions.contains(variableDS.getDimension(i).getFullName())) {
+            if (!ignoredDimensions.contains(variableDS.getDimension(i).getName())) {
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/VariableAdapter.java:496]: That's might have been an oversight indeed.
The only difference between shortName and fullName is that fullName invokes the deprecated makeFullName that will escape ".\\" chars if present in the name

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/NetCDFImageReader.java [hunk 10]
```diff
@@ -632,7 +628,10 @@ public VariableAdapter getCoverageDescriptor(Name name) {
 
     /** @see javax.imageio.ImageReader#read(int, javax.imageio.ImageReadParam) */
     @Override
-    @SuppressWarnings("PMD.ReplaceHashtableWithMap") // needed for BufferedImageConstructor
+    @SuppressWarnings({
+        "deprecation",
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/NetCDFImageReader.java:632]: Still a Dimension.
Moreover, also the setRange has been deprecated.

File: modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/CoordinateVariableTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/CoordinateVariableTest.java:66]: This is however a wrapping class that overrides the getUnit for tests.

File: modules/plugin/coverage-multidim/grib/src/test/java/org/geotools/coverage/io/grib/GribTest.java [hunk 11]
```diff
@@ -339,13 +340,13 @@ public void testGribImageWithTimeDimension() throws MalformedURLException, IOExc
 
             GeneralParameterValue[] values = {gg, time, height};
 
-            // Read with 0th date
+            // Read with 1th date
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/grib/src/test/java/org/geotools/coverage/io/grib/GribTest.java:343]: Done

File: modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/NetCDFGeoreferenceManager.java [hunk 12]
```diff
@@ -580,6 +580,7 @@ public NetCDFGeoreferenceManager(NetcdfDataset dataset) {
      * Parse the CoordinateAxes of the dataset and setup proper {@link CoordinateVariable} instances
      * on top of it and proper mapping between NetCDF dimensions and related coordinate variables.
      */
+    @SuppressWarnings("deprecation") // Missing alternative for setAxisType
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/imageio/netcdf/NetCDFGeoreferenceManager.java:583]: Right. I will use that approach

**Comment by dromagnoli** [comment]
All feedbacks have been addressed and a section in the upgrade doc has been added

File: docs/user/welcome/upgrade.rst [hunk 13]
```diff
@@ -29,6 +29,21 @@ The first step to upgrade: change the ``geotools.version`` of your dependencies
         ....
     </dependencies>
 
+.. _update32:
+
+GeoTools 32.x
+-------------
+
+NetCDF Version upgrade to 5.5.3
```
Comment (aaime) [inline docs/user/welcome/upgrade.rst:37]: The documentation should indicate which types of changes and what to do as a GeoTools user:
- Variable names
- Time interpretation
- The need to wipe the index files in the hidden directory and eventually the mosaic configurations and so on.

Could be copied over from the GeoServer docs and adapted (or linked, but maybe the GeoServer docs are interleaving information that's GS specific with the file management bits).

Comment (dromagnoli) [reply docs/user/welcome/upgrade.rst:37]: Done

**Pull request merged**